### ISSUE TITLE
[Bug]: Map Loading Gets Stuck 

### ISSUE THREAD

**Issue by SterLovesFood** [RPTools/maptool#3853]
### Describe the Bug

My players loading any map is becoming a big issue for me while trying to stream a live show.

It is normal for all 5 of my players to have to close Maptool 2-3 times and reconnect to my server in order to finish loading any map.

- At some point loading asset progress gives up and loading will no longer resume. (19/20) etc.
- Waiting a very long time will not cause the loading to begin again.
- This can happen while loading very small files according to their transfer window. (Have seen progress stop on 150kb file size)
- I have looked up the asset ID of stalled files and notice no correlation or anything odd about them.
- The largest files being loaded are around 1mb. 
- Once loading gets stuck it will not be able to resume until the player has closed map tool and reconnected.
- Even though it is stuck I can have the player start loading a new map and that will begin downloading. But also get stuck.
- I have had players clear their asset cache (based on old forum post suggestions)
- They have modified maptool.cfg (-Xmx8G -Xms8G-Xss12M).
- This still happens while only 1 player is connecting at a time.
- Can still happen when loading maps that do not have very many assets
- I tried deleting all OnCampaignLoad lib tokens as a precaution for initial connection.
- This happens with or without a repository.
- All the players have reliable stable internet connections.
- I have a computer on LAN that still runs into this problem when connecting (seems less frequent as loading speeds are faster and thus less likely to 'timeout' during a map load, but will still inevitably halt). 
- I am port forwarded as the host using TCP with all firewalls turned off.
- I know other people with the same issue so it should not be localized to just me
-  forum posts about this going back years. ( 2020 https://forums.rptools.net/viewtopic.php?t=28628 2009 https://forums.rptools.net/viewtopic.php?t=8760
- This issue is very consistent and will happen on basically every change to a new map.

### To Reproduce

1. Load a map

### Expected Behaviour

Ability for users to load a map without having to close the program. 

I have seen this issue with other friends across multiple maptool versions and old forum posts lead me to believe it has existed for some time. Some measure to automatically restart the download when it becomes stuck seems necessary. Even if current progress is thrown out, the cache is successfully being built by the files that were already loaded. Players are always able to successfully load the maps once they have everything cached. 

### Screenshots

![LoadingTransit](https://user-images.githubusercontent.com/45277765/223010811-29673fbb-cb33-4c75-a728-971038ac7ed7.png)

### MapTool Info

v1.12.2

### Desktop

All users are Windows

**Comment by emmebi** [comment]
I can confirm the bug, for 1.12.2; I don't remember if this affected us before 1.12.x, but I have definitively seen it in 1.12.2.

I will see if this week I can reproduce the bug with a large campaign and a client with an empty cache.

**Comment by Azhrei** [comment]
Since it's happening when you use a repository (but you don't say if the repo contains ALL of the image assets or just a subset), I would guess that it's a client-side problem.

There was a bug in asset cache handling fixed not too long ago — I do't know if the fix is included in 1.12.2 or if it's part of the next release.  @Phergus would likely know...  Depending on his answer, we can determine what the next step would be.  (As @FullBleed says, we don't get a lot of these reports or we'd be hearing about this much more.)

**Comment by SterLovesFood** [comment]
> Since it's happening when you use a repository (but you don't say if the repo contains ALL of the image assets or just a subset), I would guess that it's a client-side problem.

Repository has all the assets. I was exporting a new repo every week before sessions to make sure

> Have you tried running a server with 1.11.5?
I can remember running into this issue as far back as 3-4 years ago (my players treat this as normal maptool behavior and are used to having to do this - coming from a different server host who had the same problem) , but it may vary on how often it happens depending on the version of maptool.
I think it was also happening in 1.11.5, maybe less, but I can try and see if there is any difference

**Comment by emmebi** [comment]
I am consistently observing this behavior on a campaign I am using to develop adds-on. The campaign has many assets (200+) some of them quite large (15MB+).

Both the client and the server are on my LAN, connected via Ethernet. The actual campaign is on a shared disk on a Samba server.

The server computer is Windows 10, the client computer is Linux.

To reproduce the behavior, I am doing the following:
1. start MapTool on the server
2. load the campaign
3. start the server
4. wait for the map to refresh
5. clean-up the assetcache directory on the client
6. launch the MapTool on the client
7. connect the client

The client always get stuck at some point during the download. Stopping and restarting MapTool (sometimes more than once) will eventually solve the problem.

I am attaching the logs (client side, with `-d`). I have noticed that when the download get stuck, I see the following messages:

```
2023-03-18 18:05:59.208 INFO  net.rptools.maptool.client.ClientMessageHandler - emmebi-player: p got: UPDATE_ASSET_TRANSFER_MSG
2023-03-18 18:05:59.209 ERROR net.rptools.maptool.client.ClientMessageHandler - java.lang.IllegalArgumentException: Not expecting chunk: 5a3ccb918d5721e9b4346cc705c9e303
```

@SterLovesFood could you try to connect your local client to the server using the debug mode, and check if you see the same error in the logs?

**Comment by SterLovesFood** [comment]
> 
> @SterLovesFood could you try to connect your local client to the server using the debug mode, and check if you see the same error in the logs? 

Maptool is freezing if I have the log window open while connecting to the server so I'm not able to see the moment those messages would pop up. If I open it after the map loading has stalled I just have this message once every second:
![autosave](https://user-images.githubusercontent.com/45277765/226425012-9b4be599-e1f5-4f1b-8135-719e993d4dbd.png)
Which is odd because I've made sure to set the chatlog autosave and campaign autosave to 99 on both the client and host.

I've been busy and unable to test on v1.11.5

**Comment by emmebi** [comment]
You can avoid opening the log window, and get the log from the log directory:
* Windows: C:\Users\<your-user-name>\.maptool-rptools\logs\maptool.log
* Linux: /home/<your-user-name>/.maptool-rptools/logs/maptool.log

The logs themselves are text files which you can examine with any text editor.
Keep in mind that logs in debug mode are quite verbose, so ensure that you don't have information you don't want to share logged there.

**Comment by SterLovesFood** [comment]
> could you try to connect your local client to the server using the debug mode, and check if you see the same error in the logs?

I do have that in my logs

`Ash: p got: UPDATE_ASSET_TRANSFER_MSG
java.lang.IllegalArgumentException: Not expecting chunk: 5b9399d8a2cd1cccfe540029a9bd8e3f
client.ClientMessageHandler - Ash: p got: UPDATE_ASSET_TRANSFER_MSG
util.ImageManager - Asset available: eba3863e49cf5ff80b9110c87fc8e408
util.ImageManager - Loading asset: eba3863e49cf5ff80b9110c87fc8e408
util.ImageManager - BackgroundImageLoader.run(Gun,data, eba3863e49cf5ff80b9110c87fc8e408): looks like data and skipped
util.ImageManager - Notifying observers of image availability: eba3863e49cf5ff80b9110c87fc8e408
`
Though it doesnt seem to appear only when getting stuck, as those logs imply more was loaded after what is shown there.

At the bottom of my log where it did get stuck I have

'net.rptools.maptool.util.ImageManager - Asset available: 9c38f3ef429b0191fc108f432a8bd2fb
net.rptools.maptool.util.ImageManager - Loading asset: 9c38f3ef429b0191fc108f432a8bd2fb
net.rptools.maptool.util.ImageManager - Asset available: 9c38f3ef429b0191fc108f432a8bd2fb
net.rptools.maptool.util.ImageManager - Notifying observers of image availability: 9c38f3ef429b0191fc108f432a8bd2fb
net.rptools.maptool.util.ImageManager - Loading asset: 9c38f3ef429b0191fc108f432a8bd2fb
net.rptools.maptool.util.ImageManager - Image wasn't in transit: 9c38f3ef429b0191fc108f432a8bd2fb'

**Comment by sauliuskarmanovas** [comment]
This is my first time looking into this repo and I don't have the tools to build this project or a campaign to test if this fixes it but I believe I found the issue / solution. I've added a my detailed reasoning since I spent only a few hours looking at the code and I'm not familiar with nuances of this code.

Each image gets sent to the client using two message types `START_ASSET_TRANSFER_MSG` and `UPDATE_ASSET_TRANSFER_MSG`. The client needs to receive START before all the UPDATEs otherwise it throws the `java.lang.IllegalArgumentException: Not expecting chunk {ChunkId}` error. I'm not 100% positive but I believe that this would cause the client to never finish loading the map.

Why does the START reach the client before UPDATE when they are sent in the correct order? 
START message is sent by the main thread and then it passes all of the content to the AssetTransferManager to be sent by AssetProducerThread that splits the content into separate UPDATE messages and send them. The issue is that SocketClientConnection uses multiple queues for sending messages and tries to send one message from each queue before moving to the next queue.

START message is sent via the main queue (`channel == null`) but UPDATE messages use a separate one (`channel == MapToolConstants.Channel.IMAGE`). If the main queue has unsent messages that the SocketClientConnection.SenderThread needs to get through before dispatching START and the image queue is empty then the UPDATE will be sent before the START.

**Solution: send `START_ASSET_TRANSFER_MSG` via the same channel as the `UPDATE_ASSET_TRANSFER_MSG` i.e. change this line:**
https://github.com/RPTools/maptool/blob/a838571aa1d2c5f87c4d8582cdf0fbb6e6060329/src/main/java/net/rptools/maptool/server/ServerMessageHandler.java#L600
to `.sendMessage(id, MapToolConstants.Channel.IMAGE, Message.newBuilder().setStartAssetTransferMsg(msg).build());`
